File: Data Structures & Algorithms/partition-equal-subset-sum/submission-1.cpp
class Solution {
public:
    bool req(vector<int>& nums, int pos, int sum){
        if(sum==0){
            return true;
        }
        if(pos==nums.size() or sum<0){
            return false;
        }


        for(int i=pos;i<nums.size();i++){
            if(req(nums,i+1,sum-nums[i])){
                return true;
            }
        }

        // if(sum-nums[pos]<0){
        //     return req(nums,pos+1,sum);
        // }
        // else{
        //     return req(nums,pos+1,sum) or req(nums,pos+1,sum-nums[pos]);
        // }
    }

    bool canPartition(vector<int>& nums) {
        int sum = 0;
        for(int i=0;i<nums.size();i++){
            sum+=nums[i];
        }
        if(sum%2==1){
            return false;
        }

        return req(nums,0,sum/2);

    }
};

File: Data Structures & Algorithms/combination-target-sum-ii/submission-1.py
class Solution:
    def combinationSum2(self, nums: List[int], target: int) -> List[List[int]]:

        ans = []
        curr = []

        def req(i, target):
            if target == 0:
                ans.append(curr.copy())
                return

            if target < 0 or i == len(nums):
                return

            # Don't take nums[i]
            curr.append(nums[i])
            req(i+1, target - nums[i])
            curr.pop()
            while i+1<len(nums) and nums[i+1]==nums[i]:
                i = i+1
            req(i + 1, target)

            # Take nums[i]
            
        nums.sort()
        req(0, target)

        return ans
        
        